### PR TITLE
check version incompatibility

### PR DESCRIPTION
**Pull request by bn222**
When nested CNI version is 0.3.1 or less while the top level CNI version
is 0.4.0, error out. Version 0.3.1 or less doesn't support the CHECK
command. This closes issue #737.

### CODE REVIEW

**Comment by dougbtv** [comment]
Hey @bn222 ! Thanks for the contribution.

We'll also need to take care of this in the next generation configuration generation, which can be found approximately here:

https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/cmd/controller/main.go
https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/cmd/config-generation/kubeconfig.go

**Comment by bn222** [comment]
Hi @dougbtv, I'm not sure what I should change in the two files you mention. I guess you might actually mean the following file:

https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/pkg/config/generator.go#L37

I could enforce checking that the delegate configurations are using a compatible CNI version, although that would require me to change the type of Delegates in MultusConf if we want to avoid converting to json and checking the cni version in json.

Let me know what you think!

**Comment by dougbtv** [comment]
Hey Balazs, what I mean is that we have two methods for entrypoint as we're in a transition (which will be superceded by the go code, making the bash script obsolete)

So, in two different modes, either the bash runs, or the go runs (the files I linked)

Meaning that you'll have to implement the equivalent logic in the the go code as well as the bash :+1:

File: pkg/config/generator.go [hunk 1]
```diff
@@ -70,6 +73,39 @@ func NewMultusConfig(pluginName string, cniVersion string, kubeconfig string, co
...
+			delegateVersion, ok := delegatesMap["cniVersion"].(string)
+			if !ok {
+				s := "Couldn't get cni version of delegate"
+				return "", errors.New(s)
+			}
+			v, err := semver.Make(delegateVersion)
+			if err != nil {
+				return "", err
+			}
+			if v.LTE(v031) {
```
Comment (maiqueb) [inline pkg/config/generator.go:102]: I don't understand why 0.3.1 is special here. Shouldn't we just reject *any* delegates whose version is lower than 0.4.0 ?

File: pkg/config/generator.go [hunk 2]
```diff
@@ -70,6 +73,39 @@ func NewMultusConfig(pluginName string, cniVersion string, kubeconfig string, co
 // Generate generates the multus configuration from whatever state is currently
 // held
 func (mc *MultusConf) Generate() (string, error) {
+	versionFmt := "delegate cni version is %s while top level cni version is %s"
+	v040, _ := semver.Make("0.4.0")
+	v031, _ := semver.Make("0.3.1")
+	topLevel, err := semver.Make(mc.CNIVersion)
```
Comment (maiqueb) [inline pkg/config/generator.go:79]: please name this `multusCNIVersion` or something alike. 

I don't expect a version number to be in a variable named `topLevel` .

File: pkg/config/generator.go [hunk 3]
```diff
@@ -70,6 +73,39 @@ func NewMultusConfig(pluginName string, cniVersion string, kubeconfig string, co
...
+			if err != nil {
+				return "", err
+			}
+			if v.LTE(v031) {
+				s := fmt.Sprintf(versionFmt, delegateVersion, mc.CNIVersion)
+				return "", errors.New(s)
+			}
+		}
+	}
+
```
Comment (maiqueb) [inline pkg/config/generator.go:108]: Please encapsulate this in a separate func / method - the simple `Generate` method turned into 95% checking a corner case and 5% generating.

I'd actually call this new function from the constructor, thus preventing the creation of an invalid config.

You can change the signature of the constructor to also return an error.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Other than the comments I'm providing in the code, I can't help to notice that the new functionality  is not being unit tested; you'll have to provide these tests as well.

I also disagree with your choice of where to implement this: I think it should be done in the configuration generator constructor, not in its generate method.

File: pkg/config/generator.go (hunk 1 above)
Comment (bn222) [reply pkg/config/generator.go]: I addressed this by only checking nested version < 0.4.0 when top level version >= 0.4.0.

File: pkg/config/generator.go (hunk 3 above)
Comment (bn222) [reply pkg/config/generator.go]: Done. This caused a few more changes elsewhere as well as the error is propagated.

File: pkg/config/generator_test.go [hunk 4]
```diff
@@ -45,24 +45,28 @@ var primaryCNIConfig = map[string]interface{}{
...
+	if err != nil {
+		return multusConfig, err
+	}
+	err = multusConfig.Mutate(withDelegates(primaryCNIPluginConfig.(map[string]interface{})))
+	return multusConfig, err
 }
 
 func TestBasicMultusConfig(t *testing.T) {
-	multusConfig := newMultusConfigWithDelegates(
+	multusConfig, _ := newMultusConfigWithDelegates(
```
Comment (maiqueb) [inline pkg/config/generator_test.go:58]: Now that the method returns an error, it feels weird to swallow it up.

Do you mind asserting it does not error on the tests ? Or better; provide another helper for it ?

**Review by maiqueb — CHANGES_REQUESTED** [review]
It feels a bit weird to not check the error on `newMultusConfigWithDelegates` now that it can return one. 

Unfortunately, I don't have a good alternative to doing that: either check it on the test's body, or introduce another helper for it.

Pretty much everything else is opinionated.

**Review by maiqueb — APPROVED** [review]
Thanks for the patience, and for the patch.

@dougbtv I've **only** reviewed the golang related code.

**Comment by s1061123** [comment]
It looks good to me. Apologize that the PR takes a very long time. Let me merge into master. Thank you so much!

**Pull request merged**